futures-util/src/stream/forward.rs

```diff
@@ -1,112 +1,94 @@
+use core::marker::Unpin;
 use core::mem::PinMut;
 
-use futures_core::{Future, Poll, Stream};
-use futures_core::task;
+use futures_core::{task, Future, Poll, Stream};
+use futures_sink::Sink;
 
 use stream::{StreamExt, Fuse};
 
+const INVALID_POLL: &str = "polled `Forward` after completion";
+
 /// Future for the `Stream::forward` combinator, which sends a stream of values
-/// to a sink and then waits until the sink has fully flushed those values.
+/// to a sink and then flushes the sink.
+///
+/// Note: this is only usable with `Unpin` sinks, so `Sink`s that aren't `Unpin`
+/// will need to be pinned in order to be used with this combinator.
```

Comment:
- It would be good if the concrete reason was given. Maybe in parenthesis.
- I think you men "...will need to be ~~pinned~~ boxed in order..."

Reply:
> I think you mean "...will need to be pinned boxed in order..."

I meant that they need to be pinned. This can be done by putting them in a `PInBox`, but it's not the only way.

Reply:
Ah I get now what you mean. Maybe it's just me, but I think it is really hard to understand what is meant by this sentence. It would be clearer if it was expanded a bit. "e.g. by putting them in `PinBox`es or by pinning them on the stack and giving `PinMut` references to `forward` instead"

Reply:
Ok. Then I suppose you could remove "so `Sink`s that aren't `Unpin` will need to be pinned in order to be used with this combinator". Without this it's much clearer.

My main gripe with that sentence is that it kinda sounds like "Pinning makes a type `Unpin`". But that's not true of course, instead the type that we get by pinning it will be `Unpin`.

Reply:
Pinning results in a value that is `Unpin`, yeah. The naming works out to be sort of funny.